### ISSUE TITLE
Whisper Alignment Heads calculation for custom model

### ISSUE THREAD

**Issue by DavraYoung** [huggingface/transformers#28365]
### Feature request

Token timestamps work great on the pretrained model, but once the model is finetuned, token timestamp are no more correct.
I tried to dig deeper and found that the token timestamps calculation is done using cross attention heads with some dynamic time warping calculation over token sequence.

it would be good to have some discussion or script that allow to select the best alignment heads for any whisper model.

What were your approach for selecting those alignment heads?

### Motivation

Token timestamps, long form transcription, hallucination detection etc.

### Your contribution

I have tried calculating best alignment heads using this approach:

0. Select 100 good enough sample audios.
1. Use some CTC model for per word timestamp calculation, calculate per word timestamps
2. Run whisper with token timestamps and get timestamps considering each attention head separately
3. Consider the time difference for each attention head with the one from CTC model (for matching token transcriptions)
4. Select the attention head with best accuracy
5. Repeat from step 1, using the selected head, calculate all other heads that in combination with current heads can improve the accuracy

I was able to achieve average 200ms accuracy for token timestamps. Accuracy on best 80% is 40ms. The model fails only on some cases having 2000-4000ms shift, which can be detected using some scripts.

**Comment by DavraYoung** [comment]
@sanchit-gandhi hi, have you tried this with distilled model?

**Comment by ArthurZucker** [comment]
Hey 🤗  thanks for opening an issue! We try to keep the github issues for bugs/feature requests.
Could you ask your question on the [forum](https://discuss.huggingface.co/) instead? I'm sure the community will be of help!

Thanks!

**Comment by sanchit-gandhi** [comment]
Hey @DavraYoung! Indeed - as @ArthurZucker mentioned the forum is the best place to ask such questions. Linking some resources that may be of help for reference for your question to get you going:
* Answer from Whisper author: https://github.com/openai/whisper/discussions/1388#discussioncomment-7568428
* Thread on the HF Hub: https://huggingface.co/distil-whisper/distil-small.en/discussions/7

**Comment by LaurinmyReha** [comment]
checkout this repo for the best individual alignment heads of the large v2 and v3 variants.

https://github.com/nyrahealth/CrisperWhisper/blob/develop/run_experiments/experiments/head_results.json

The Repo also contains useful code to calculate these head results for other models as long as you are in the posession of a dataset with high quality timestamps like timit ( https://paperswithcode.com/dataset/timit). 


https://github.com/nyrahealth/CrisperWhisper
